plugin/serve.go

```diff
@@ -84,65 +104,127 @@ func Serve(opts *ServeOpts) {
 		log.SetOutput(logger.StandardWriter(&hclog.StandardLoggerOptions{InferLevels: true}))
 	}
 
-	// since the plugins may not yet be aware of the new protocol, we
-	// automatically wrap the plugins in the grpc shims.
-	if opts.GRPCProviderFunc == nil && opts.ProviderFunc != nil {
+	if opts.ProviderAddr == "" {
+		opts.ProviderAddr = "provider"
+	}
+
+	var err error
+
+	switch {
+	case opts.ProviderFunc != nil && opts.GRPCProviderFunc == nil:
 		opts.GRPCProviderFunc = func() tfprotov5.ProviderServer {
 			return schema.NewGRPCProviderServer(opts.ProviderFunc())
 		}
+		err = tf5serverServe(opts)
+	case opts.GRPCProviderFunc != nil:
+		err = tf5serverServe(opts)
+	case opts.GRPCProviderV6Func != nil:
+		err = tf6serverServe(opts)
+	default:
+		err = errors.New("no provider server defined in ServeOpts")
 	}
 
-	serveConfig := plugin.ServeConfig{
-		HandshakeConfig: Handshake,
-		GRPCServer: func(opts []grpc.ServerOption) *grpc.Server {
-			return grpc.NewServer(opts...)
-		},
-		Logger: opts.Logger,
-		Test:   opts.TestConfig,
+	if err != nil {
+		log.Printf("[ERROR] Error starting provider: %s", err)
```

Comment:
Same comment about the `.Fatal` here. If this fails, definitely the provider isn't working (and will `exit(0)`).